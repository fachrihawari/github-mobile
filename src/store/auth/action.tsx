import { FETCH_USER_REQUEST, LOGOUT } from "./constant";
import { createAction } from "../helpers";

export const fetchUserRequest = (username: string, password: string, OTP: string='') =>
  createAction(FETCH_USER_REQUEST, { username, password, OTP });

export const logout = () => createAction(LOGOUT);
