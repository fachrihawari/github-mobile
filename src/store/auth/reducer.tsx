import { AnyAction } from "redux";
import {
  FETCH_USER_BEGIN,
  FETCH_USER_SUCCESS,
  FETCH_USER_FAILURE
} from "./constant";
import { transformProfile } from "./transform";

export interface IUser {
  id: number;
  name: string;
  email: string;
  avatarUrl: string;
  company: string;
  location: string;
  blog: string;
  bio: string;
  publicRepos: number;
  publicGists: number;
  followers: number;
  following: number;
}

export interface IState {
  isLoggedIn: boolean;
  isLoading: boolean;
  error: string | null;
  user: IUser | null;
}

const initialState: IState = {
  isLoggedIn: false,
  isLoading: false,
  error: null,
  user: null
};

export default (state: IState = initialState, action: AnyAction): IState => {
  switch (action.type) {
    case FETCH_USER_BEGIN:
      return {
        ...state,
        isLoading: true,
        error: null
      };
    case FETCH_USER_SUCCESS:
      return {
        ...state,
        isLoading: false,
        isLoggedIn: true,
        user: transformProfile(action.payload.profile)
      };
    case FETCH_USER_FAILURE:
      return {
        ...state,
        isLoading: false,
        error: action.payload.error
      };
    default:
      return state;
  }
};
