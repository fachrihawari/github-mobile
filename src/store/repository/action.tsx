import { FETCH_COMMIT_REQUEST } from "./constant";
import { createAction } from "../helpers";

export const fetchCommitRequest = (
  repository: string,
  page?: number,
  perPage?: number
) => createAction(FETCH_COMMIT_REQUEST, { repository, page, perPage });
