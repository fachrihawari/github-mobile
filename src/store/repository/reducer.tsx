import { AnyAction } from "redux";
import {
  FETCH_COMMIT_BEGIN,
  FETCH_COMMIT_SUCCESS,
  FETCH_COMMIT_FAILURE
} from "./constant";
import { transformCommit } from "./transform";

export interface ICommit {
  sha: string;
  date: Date;
  message: string;
  authorName: string;
  authorEmail: string;
  authorUsername: string;
  authorAvatarUrl: string;
  authorUrl: string;
}

export interface IState {
  isLoading: boolean;
  error: string | null;
  commits: Array<ICommit>;
  page: number;
  perPage: number;
}

const initialState: IState = {
  isLoading: false,
  error: null,
  page: 1,
  perPage: 10,
  commits: []
};

export default (state: IState = initialState, action: AnyAction): IState => {
  switch (action.type) {
    case FETCH_COMMIT_BEGIN:
      return {
        ...state,
        isLoading: true,
        error: null
      };
    case FETCH_COMMIT_SUCCESS:
      return {
        ...state,
        isLoading: false,
        page: action.payload.currentPage,
        commits: action.payload.commits.map(transformCommit)
      };
    case FETCH_COMMIT_FAILURE:
      return {
        ...state,
        isLoading: false,
        error: action.payload.error
      };
    default:
      return state;
  }
};
