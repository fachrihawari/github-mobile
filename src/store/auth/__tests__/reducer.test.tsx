import reducer, { initialState, IState } from '../reducer'
import { AnyAction } from 'redux';
import { logout, fetchUserBegin, fetchUserSuccess, fetchUserFailure } from '../action';
import { transformProfile } from '../transform';

describe('reducer', () => {

  let currentState: IState

  function setup() {
    currentState = Object.assign({}, initialState)
  }

  beforeEach(setup)

  it('should return the initial state', () => {
    expect(reducer(undefined, {} as AnyAction)).toEqual(currentState)
    expect(reducer(undefined, logout())).toEqual(currentState)
  })

  it('should return the begin state', () => {
    currentState.isLoading = true
    expect(reducer(undefined, fetchUserBegin())).toEqual(currentState)
  })

  it('should return the success state', () => {
    const profile = {
      "login": "github",
      "id": 9919,
      "node_id": "MDEyOk9yZ2FuaXphdGlvbjk5MTk=",
      "avatar_url": "https://avatars1.githubusercontent.com/u/9919?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/github",
      "html_url": "https://github.com/github",
      "followers_url": "https://api.github.com/users/github/followers",
      "following_url": "https://api.github.com/users/github/following{/other_user}",
      "gists_url": "https://api.github.com/users/github/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/github/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/github/subscriptions",
      "organizations_url": "https://api.github.com/users/github/orgs",
      "repos_url": "https://api.github.com/users/github/repos",
      "events_url": "https://api.github.com/users/github/events{/privacy}",
      "received_events_url": "https://api.github.com/users/github/received_events",
      "type": "Organization",
      "site_admin": false,
      "name": "GitHub",
      "company": null,
      "blog": "https://github.com/about",
      "location": "San Francisco, CA",
      "email": "support@github.com",
      "hireable": null,
      "bio": "How people build software.",
      "public_repos": 306,
      "public_gists": 0,
      "followers": 0,
      "following": 0,
      "created_at": "2008-05-11T04:37:31Z",
      "updated_at": "2019-07-15T14:26:15Z"
    }

    currentState.isLoading = false
    currentState.isLoggedIn = true
    currentState.user = transformProfile(profile)
    
    expect(reducer(undefined, fetchUserSuccess(profile))).toEqual(currentState)
  })

  it('should return the failure state', () => {
    const error = 'Bad Credentials'
    currentState.error = error
    expect(reducer(undefined, fetchUserFailure(error, false))).toEqual(currentState)
  })
})