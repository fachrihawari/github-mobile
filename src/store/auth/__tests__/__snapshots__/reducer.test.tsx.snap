// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer should return the begin state 1`] = `
Object {
  "error": null,
  "isLoading": true,
  "isLoggedIn": false,
  "needOTP": false,
  "user": null,
}
`;

exports[`reducer should return the failure state 1`] = `
Object {
  "error": "Bad Credentials",
  "isLoading": false,
  "isLoggedIn": false,
  "needOTP": false,
  "user": null,
}
`;

exports[`reducer should return the initial state 1`] = `
Object {
  "error": null,
  "isLoading": false,
  "isLoggedIn": false,
  "needOTP": false,
  "user": null,
}
`;

exports[`reducer should return the success state 1`] = `
Object {
  "error": null,
  "isLoading": false,
  "isLoggedIn": true,
  "needOTP": false,
  "user": Object {
    "avatarUrl": "https://avatars1.githubusercontent.com/u/9919?v=4",
    "bio": "How people build software.",
    "blog": "https://github.com/about",
    "company": null,
    "email": "support@github.com",
    "followers": 0,
    "following": 0,
    "id": 9919,
    "location": "San Francisco, CA",
    "name": "GitHub",
    "publicGists": 0,
    "publicRepos": 306,
  },
}
`;
